### PR TITLE
Translation fixes for `{:extern}` when used for exports and for `--test-assumptions=externs`

### PR DESCRIPTION
**Pull request by keyboardDrummer**
### Description
- Fix incorrect translated code when using `{:extern}` to export members
- Fix incorrect translated code when using `--test-assumptions=externs`

### How has this been tested?
- Add a CLI test

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/IntegrationTests/TestFiles/LitTests/LitTest/comp/extern/dfyconfig.toml
Comment (robin-aws) [inline Source/IntegrationTests/TestFiles/LitTests/LitTest/comp/extern/dfyconfig.toml:1]: Nice use of project files for tests :)

File: Source/DafnyCore/Rewriters/ExpectContracts.cs [hunk 1]
```diff
@@ -119,10 +117,9 @@ private MemberDecl GenerateFunctionWrapper(TopLevelDeclWithMembers parent, Membe
 
     newFunc.Body = callExpr;
 
-    var localName = origFunc.Result?.Name ?? "__result";
-    var localExpr = new IdentifierExpr(tok, localName) {
-      Type = newFunc.ResultType
-    };
+    var resultVar = origFunc.Result ?? new Formal(tok, "#result", origFunc.ResultType, false, false, null);
+    var localName = origFunc.Result?.Name ?? "__BLA__result";
```
Comment (robin-aws) [inline Source/DafnyCore/Rewriters/ExpectContracts.cs:121]: BLA?

File: Source/DafnyCore/Backends/Java/JavaCodeGenerator.cs [hunk 2]
```diff
@@ -884,7 +884,7 @@ protected override string TypeName_UDT(string fullCompileName, List<TypeParamete
     // We write an extern class as a base class that the actual extern class
     // needs to extend, so the extern methods and functions need to be abstract
     // in the base class
-    protected override bool IncludeExternMembers { get => true; }
+    protected override bool IncludeImportedMembers => true;
```
Comment (robin-aws) [inline Source/DafnyCore/Backends/Java/JavaCodeGenerator.cs:887]: Consider using "ExternImported" rather than just "Imported" throughout, only because if we also start saying just "Exported" for the other direction, it will be easily confused with export sets.

File: Source/IntegrationTests/TestFiles/LitTests/LitTest/comp/extern/exportedAndTestExternAssumptions.dfy
Comment (robin-aws) [inline Source/IntegrationTests/TestFiles/LitTests/LitTest/comp/extern/exportedAndTestExternAssumptions.dfy:1]: Worth a comment what's actually being tested here since it's not obvious. I think it's the fact that it previously didn't work because some extern-exported Dafny bodies were dropped.

I think you're also not actually testing the --test-assumptions fixes AFAICT? At a minimum you'd want to have a Dafny call to Builder that doesn't satisfy Pre.

**Review by robin-aws — COMMENTED** [review]
Great improvement, just some style/doc suggestions (and remove all the commented out code please :)

File: Source/DafnyCore/Rewriters/ExpectContracts.cs (hunk 1 above)
Comment (robin-aws) [reply Source/DafnyCore/Rewriters/ExpectContracts.cs]: Well it's known to at least one more person now, so thank you for that :)

File: Source/DafnyCore/Backends/SinglePassCodeGenerator/SinglePassCodeGenerator.cs [hunk 3]
```diff
@@ -1529,15 +1529,28 @@ public void Compile(Program program, ConcreteSyntaxTree wrx) {
       EmitFooter(program, wrx);
     }
 
-    protected (bool classIsExtern, bool included) GetIsExternAndIncluded(ClassLikeDecl cl) {
+
+    public bool IsImported(IAttributeBearingDeclaration me) {
```
Comment (robin-aws) [inline Source/DafnyCore/Backends/SinglePassCodeGenerator/SinglePassCodeGenerator.cs:1533]: Would prefer `IsExternallyImported` since you made the other change, but could be changed in a future PR

File: Source/DafnyCore/Rewriters/ExpectContracts.cs [hunk 4]
```diff
@@ -119,49 +117,28 @@ private MemberDecl GenerateFunctionWrapper(TopLevelDeclWithMembers parent, Membe
...
-      var locs = new List<LocalVariable> { local };
-      var varDeclStmt = new VarDeclStmt(decl.RangeToken, locs, new AssignStatement(decl.RangeToken, lhss, rhss) {
-        ResolvedStatements = new List<Statement>() { assignStmt }
-      });
-      localExpr.Var = local;
-      callStmt = varDeclStmt;
-    } else {
-      localExpr.Var = origFunc.Result;
-      callStmt = assignStmt;
-    }
```
Comment (robin-aws) [inline Source/DafnyCore/Rewriters/ExpectContracts.cs:146]: Not completely following why this logic became unnecessary?

File: Source/DafnyCore/Rewriters/ExpectContracts.cs [hunk 5]
```diff
@@ -178,21 +155,22 @@ private MemberDecl GenerateMethodWrapper(TopLevelDeclWithMembers parent, MemberD
...
+    return newMethod;
   }
 
 
-  private void RegisterResolvedByMethod(Function f, TopLevelDeclWithMembers cl) {
+  private void RegisterResolvedByMethod(Formal resultVar, Function f, TopLevelDeclWithMembers cl) {
 
     var tok = f.ByMethodTok;
-    var resultVar = f.Result ?? new Formal(tok, "#result", f.ResultType, false, false, null);
+    // var resultVar = f.Result ?? new Formal(tok, "#result", f.ResultType, false, false, null);
```
Comment (robin-aws) [inline Source/DafnyCore/Rewriters/ExpectContracts.cs:165]:
```suggestion
```

File: Source/TestDafny/MultiBackendTest.cs [hunk 6]
```diff
@@ -22,6 +22,9 @@ public class ForEachCompilerOptions {
   [Value(1, MetaName = "Dafny CLI arguments", HelpText = "Any arguments following '--' will be passed to the dafny CLI unaltered.")]
   public IEnumerable<string> OtherArgs { get; set; } = Array.Empty<string>();
 
+  [Option("run-fails", HelpText = "Whether the running program should return a non-zero exit code")]
+  public bool RunShouldFail { get; set; } = false;
```
Comment (robin-aws) [inline Source/TestDafny/MultiBackendTest.cs:26]: Good feature to add, but would strongly prefer `--expect-exit-code` instead, since we already have that option on `for-each-resolver`

**Review by robin-aws — CHANGES_REQUESTED** [review]

File: Source/DafnyCore/Rewriters/ExpectContracts.cs (hunk 4 above)
Comment (keyboardDrummer) [reply Source/DafnyCore/Rewriters/ExpectContracts.cs:146]: Despite writing the previous code, I don't understand well either, but this was conditional on `origFunc.Result == null` and above I do `var resultVar = origFunc.Result ?? new Formal(tok, "#result", newFunc.ResultType, false, false, null);` which handles the same thing.

File: Source/TestDafny/MultiBackendTest.cs (hunk 6 above)
Comment (keyboardDrummer) [reply Source/TestDafny/MultiBackendTest.cs:26]: I can't pass in an exit code because I don't know what it will be. Dafny does not define what the exit code will be for programs that do not pass runtime contract checks. It could be different per target. All I know it that it won't be a success exit code.

Comment (robin-aws) [reply Source/TestDafny/MultiBackendTest.cs:26]: Ah fair! I can live with this then.

**Review by robin-aws — APPROVED** [review]
